### PR TITLE
AMQP-650: AsyncRabbitTemplate Refactoring

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/AMQP-650

- Extract `AsyncAmqpTemplate` interface (see AMQP-616)
- Accept any `AbstractMessageListenerContainer` for replies
- group all `sendAndReceive` methods together

### CODE REVIEW

File: spring-amqp/src/main/java/org/springframework/amqp/core/AsyncAmqpTemplate.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+	/**
+	 * Convert the object to a message and send it to the default exchange with the
+	 * provided routing key.
+	 * @param routingKey the routing key.
+	 * @param message the message.
+	 * @param <C> the expected result type.
+	 * @return the {@link AmqpConverterFuture}.
+	 */
+	<C> AmqpConverterFuture<C> convertSendAndReceive(String routingKey, Object message) throws AmqpException;
```
Comment (artembilan) [inline spring-amqp/src/main/java/org/springframework/amqp/core/AsyncAmqpTemplate.java]: I wonder what is a difference of this method from others that only this one is throwing `AmqpException`, although the last one is `RuntimeException` any way...

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java [hunk 2]
```diff
@@ -71,15 +74,15 @@
  * @since 1.6
  */
 public class AsyncRabbitTemplate implements SmartLifecycle, MessageListener, ReturnCallback, ConfirmCallback,
-		BeanNameAware {
+		BeanNameAware, AsyncAmqpTemplate {
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java]: Nit-pick, but `AsyncAmqpTemplate` is the main contract of this class. So, it should be first in the `implements` declaration. :smile:

File: spring-amqp/src/main/java/org/springframework/amqp/core/AsyncAmqpTemplate.java (hunk 1 above)
Comment (garyrussell) [inline spring-amqp/src/main/java/org/springframework/amqp/core/AsyncAmqpTemplate.java]: Looks like it was a copy/paste from `RabbitTemplate`; I don't remember but I must have removed them from the other methods, I am not sure the value of `throws RuntimeException` except, perhaps documentation. But they should be consistent. I can go either way - WDYT?

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java (hunk 2 above)
Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java]: True.

File: spring-amqp/src/main/java/org/springframework/amqp/core/AsyncAmqpTemplate.java (hunk 1 above)
Comment (artembilan) [inline spring-amqp/src/main/java/org/springframework/amqp/core/AsyncAmqpTemplate.java]: I would have it removed.
Thanks

**Comment by artembilan** [comment]
LGTM.
It is ready for merge, but having my problem with tests it would be better to merge #471 first.
Thanks

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-amqp/commit/73fcb2c816683d4b0aadb3d54608d6ea67c6cac9 after rebase.

**Pull request closed without merging**